Add Speed Menu

MiladNazeri opened the pull request:
- Added a speed menu similar to the Avatar Size Menu Options

Specifically for @tomscho and @ZappoMan

hifi-gustavo commented:
Can one of the admins verify this patch?

hifi-gustavo commented:
Can one of the admins verify this patch?

sethalves commented:
test this please

sethalves commented:
I think your editor isn't configured to indent as hifi's coding standard requires.  there shouldn't be any tab characters, and indents should all be 4 spaces.

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1739/HighFidelity-Beta-PR10988-da3245e8e952212e83338a4c9d00acc3d9063e1b.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1739/HighFidelity-Beta-PR10988-da3245e8e952212e83338a4c9d00acc3d9063e1b.exe

MiladNazeri commented:
My apologies, thank you for pointing out.

themelissabrown commented:
@MiladNazeri we usually have our developers submit a test plan. Can you write up a quick description of how a QA tester should test this?

AndrewMeadows commented on libraries/avatars/src/AvatarData.h:
```diff
@@ -706,6 +706,9 @@ public slots:
     virtual void maybeUpdateSessionDisplayNameFromTransport(const QString& sessionDisplayName) { } // No-op in AvatarMixer
 
+	// Body speed
+    float _speed;
```
Could you the variable name `_speed` to be a bit more descriptive, to something like `_walkSpeed`,` _targetSpeed`, or something like that?

Also, this variable is only relevant for the `MyAvatar` class, not the `AvatarData` base class.  Move it into `MyAvatar.h`.

Please initialize it with a default value in the `MyAvatar` ctor.  Declare it like this:
`
float _walkSpeed { DEFAULT_WALKING_SPEED };`

Finally, as hinted in my suggestion above, change the name of `MAX_WALKING_SPEED` to something like `DEFAULT_WALKING_SPEED`.

MiladNazeri commented:
@AndrewMeadows - Thank you for the feedback, just pushed those changes.

@themelissabrown - Thank you for pointing that out.  Here is a quick test Plan:

1.  Under the Avatar Menu->Speed - Choosing to Increase the speed should increase the Avatar's speed by 2.0 until it reaches the MAX_AVATAR_SPEED(currently 30.0).  Hitting the "]" key will shortcut to this as well.
2.  Decreasing the speed progressively decreases the Avatar's speed by 2.0 until it reaches the DEFAULT_WALKING_SPEED(currently 2.6f).  Likewise, hitting the "[" key will shortcut to this.
3. Choosing Reset speed will take the speed back to the DEFAULT_WALKING_SPEED.

AndrewMeadows commented on interface/src/avatar/MyAvatar.h:
```diff
@@ -669,6 +672,9 @@ public slots:
     float _yawSpeed; // degrees/sec
     float _pitchSpeed; // degrees/sec
 
+    const float DEFAULT_WALKING_SPEED = 2.6f;
```
Don't declare `DEFAULT_WALKING_SPEED` here inside the class because it creates a per-instance data member `myAvatarInstance.DEFAULT_WALKING_SPEED`.  Move it above the class near the top, under all the `#includes`.

AndrewMeadows commented on interface/src/avatar/MyAvatar.h:
```diff
@@ -669,6 +672,9 @@ public slots:
     float _pitchSpeed; // degrees/sec
 
+    const float DEFAULT_WALKING_SPEED = 2.6f;
+    float _walkSpeed{ DEFAULT_WALKING_SPEED };
```
Put a space after `_walkSpeed` and the `{`.

MiladNazeri commented:
@AndrewMeadows - I noticed that there are some similar declarations that may need to also be moved out of the class on lines 708, 732, and 733.   Thoughts on those?

AndrewMeadows commented on interface/src/Application.cpp:
```diff
@@ -3199,6 +3199,15 @@ void Application::keyPressEvent(QKeyEvent* event) {
+
+            case Qt::Key_BracketRight:
+                getMyAvatar()->increaseSpeed();
+                break;
```
We'll probably change or remove these key bindings eventually (soon?).  There is an effort to remove random single-key bindings that people can press by accident.  My guess is that we'll eventually want to enable walk-speed changes in a different way, but I'm not sure what that will look like.  This is a start.

AndrewMeadows commented:
Here is the summary of the changes:

Default (and minimum) walk speed is about 2.6 m/sec and max avatar speed (whether walking or flying) is 30 m/sec.

You can now increase your avatar's walk speed by 2 m/sec by pressing the `]` key (up to max avatar speed) and decrease the speed by pressing `[` (down to default walk speed).

These changes are on a per-session basis: they are not saved/loaded to/from preferences.  Each new login should reset the avatar's walk speed to default.

 The **test plan** would be to make sure all of that ^^^ above is working as described.  Also test how changing the speeds affects flying around.  What happens when the avatar speed is maxed out?  Is it still usable?

The `increaseSpeed()`, `decreaseSpeed()`, and `resetSpeed()` functions are created as SLOT's on the `MyAvatar` object so they may automatically be exposed to JS.  Perhaps make or modify a script to call `MyAvatar.increaseSpeed()` and its friends to see if they work.

MiladNazeri commented:
@AndrewMeadows 

After reviewing the coding standards, I wanted to fix how the private
variable _walkSpeed was being accessed and modified.

- Added getter and setter for walkSpeed
- Changed increaseSpeed(), decreaseSpeed(), and resetSpeed() to use the new
methods

AndrewMeadows commented:
@MiladNazeri, either that bit of the coding standard is too strict or it is unclear.  We certainly don't follow it to the letter. IMO we only need to add a setter and getter for a data member that is to be accessed from **outside** of the class.  In this case `_walkSpeed` is only ever used or changed by `MyAvatar` -- it is **internal only**.  Accessing internal data through getters and setters tends to make the code less readable (IMO) and could add function call overhead (unless we have faith the compiler is smart enough to optimize it out).  Occasionally it might make sense to add a getter and setter for an internal data member that is only used internally: for easier debugging.

I'll let you chose the path forward:

(a) Remove the getter and setter or...
(b) leave them in place but make the getter method `const`.

MiladNazeri commented:
@AndrewMeadows - Thank you for the insight and for clearing that up.  I went with the removal of the getters/setters.

AndrewMeadows commented:
This looks good.  My **DEV_APPROVED** vote still stands.

There are two test plans suggested in earlier comments (I missed the first one and suggested my own version).

YuppyPlays commented:
Testing.

YuppyPlays commented:
Testing complete:

Under the Avatar Menu->Speed - Choosing to Increase the speed should increase the Avatar's speed by 2.0 until it reaches the MAX_AVATAR_SPEED(currently 30.0). Hitting the "]" key will shortcut to this as well.
**Fail, It seems like the keystroke for "]" didn't function / do anything - Increasing the speed however did work from menu.**

Decreasing the speed progressively decreases the Avatar's speed by 2.0 until it reaches the DEFAULT_WALKING_SPEED(currently 2.6f). Likewise, hitting the "[" key will shortcut to this.
**Fail, It seems like the keystroke for "[" didn't function / do anything - Decreasing the speed however did work from menu.**

Choosing Reset speed will take the speed back to the DEFAULT_WALKING_SPEED.
**Pass**

Please let me know if there is anything else I can further do to assist :)

MiladNazeri commented:
@YuppyPlays, 

The keyboard short cuts were actually made in a later commit, but I noticed that Gustavo compiled versions during the first commit.   I couldn't access the versions that were compiled above to test so was just curious about the testing process in general.  

Besides the addition of the keyboard shortcuts, the only thing changed was the formatting to match the coding standards a bit more, which is why the functionality was always consistent between versions.

If you don't mind clarifying the above for me a bit, as far as what was actually being tested - a build based off the last commit, or what may be a build from an older commit  made by Gustavo (so far I like Gustavo better than Siri; might be the beret) - then that might shed a little clarity on why those shortcuts didn't work.  

@AndrewMeadows mentioned getting away from single key shortcuts so maybe they could be eliminated altogether, as I was mostly attempting to keep consistency with the avatar size menu and make it easier to test instead of clicking through the menu.  

I appreciate all the help from the team on this PR so far, I learned a lot from the process for sure.

themelissabrown commented:
jenkins, build this please

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2058/HighFidelity-Beta-PR10988-fb77a96080f4346ceb383df9c47fd021bcbeaa99.dmg

marko8904 commented:
@MiladNazeri @themelissabrown Please note that we have switched the Jenkins build system to use 5.9.1 and the new Visual Studio 2017 toolchain. For the build to succeed on windows please merge master into this PR.

MiladNazeri commented:
@marko8904 Just finished updating, thank you for the notice

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2080/HighFidelity-Beta-PR10988-263a0838bd9ad55a4710cce34fa744ebff4c606d.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2080/HighFidelity-Beta-PR10988-263a0838bd9ad55a4710cce34fa744ebff4c606d.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2084/HighFidelity-Beta-PR10988-542e9241e1718eb1fb28dc10731efed9f6418f9e.dmg

themelissabrown commented:
build this please

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2264/HighFidelity-Beta-PR10988-542e9241e1718eb1fb28dc10731efed9f6418f9e.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2264/HighFidelity-Beta-PR10988-542e9241e1718eb1fb28dc10731efed9f6418f9e.exe

themelissabrown commented:
@MiladNazeri still failing on the linux build. Can you see the details or would you like me to email them to you?

MiladNazeri commented:
@themelissabrown just seeing this now and tried to check the report but wouldn't give me access.  Yes, could you please email me that?

themelissabrown commented:
build this please

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3063/HighFidelity-Beta-PR10988-542e9241e1718eb1fb28dc10731efed9f6418f9e.dmg

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3067/HighFidelity-Beta-PR10988-326a3fdce5dfe9743c533fab52142de3fd2565f0.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3067/HighFidelity-Beta-PR10988-326a3fdce5dfe9743c533fab52142de3fd2565f0.exe

tapalisa commented:
Things to note:
We have been attempting to avoid single key shortcuts.
The Oculus touch controllers have a lower speed when the joystick is less than fully engaged. How does this behavior interact with these settings?

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3072/HighFidelity-Beta-PR10988-9922239075d196908d2c37ae887c02b286dc8b7a.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3072/HighFidelity-Beta-PR10988-9922239075d196908d2c37ae887c02b286dc8b7a.dmg

ZappoMan commented:
@MiladNazeri - this PR has a couple of small merge conflicts -can you fix them?
